src/optimizer/topn_optimizer.cpp

```diff
@@ -28,14 +31,21 @@ bool TopN::CanOptimize(LogicalOperator &op) {
 		}
 
 		auto child_op = op.children[0].get();
+		if (context) {
+			// estimate child cardinality if the context is available
+			child_op->EstimateCardinality(*context);
+		}
 
-		auto constant_limit = static_cast<double>(limit.limit_val.GetConstantValue());
-		auto child_card = static_cast<double>(child_op->estimated_cardinality);
+		if (child_op->has_estimated_cardinality) {
+			// only bail if we have an estimated cardinality
+			auto constant_limit = static_cast<double>(limit.limit_val.GetConstantValue());
+			auto child_card = static_cast<double>(child_op->estimated_cardinality);
 
-		// if the child cardinality is not 98 times more than the
-		bool limit_is_large = constant_limit > 5000;
-		if (constant_limit > child_card * 0.007 && limit_is_large) {
-			return false;
+			// if the child cardinality is not 98 times more than the
```

Comment:
Seems like I never finished this comment. Looking back at the original PR https://github.com/duckdb/duckdb/pull/17140, the comment can be something like 
```
If the limit is > 0.7% of the child cardinality, sorting the whole table is faster
```